Prepare for npm publish, Upgrade rules_nodejs to 2.0.0

rzhw opened the pull request:
Publishing to npm addresses the version skew issue discussed internally at go/postcss-build-rules?disco=AAAADYbh_hE. This PR corresponds to #6, but let's not close that until we perform the publish.

Where publishing to npm is primarily useful is the usage of `peerDependencies`, such that downstream dependents are required/asked to avoid version skew of `postcss` as enforced by npm cli/yarn.

Also document installation of these build rules in the README - this partially addresses #34.

The plan is to publish this as version 0.5.0 (totally arbitrary version number - but it reflects progress made since initial publish), both on npm and GitHub such that `http_archive` can be used as well.

I've tested the npm package locally with the following steps:

1. Run `bazel build :npm_package`, then go to another Bazel workspace with rules_nodejs set up
2. Add `"@bazel/postcss": "relative/path/to/rules_postcss/bazel-bin/npm_package",` to `package.json`
3. Call `rules_postcss_dependencies` from `WORKSPACE`
4. Verify using `postcss_binary` etc. rules from `@build_bazel_rules_postcss//:defs.bzl` work as expected

If desired, I can put up my test repo (this may take time per [go/releasing](https://opensource.google/docs/releasing/)).

Here is the `npm pack` artifact created as a result of running `bazel run :npm_package.pack`: [bazel-postcss-0.0.1.tgz.zip](https://github.com/bazelbuild/rules_postcss/files/4778051/bazel-postcss-0.0.1.tgz.zip)

rzhw committed:
Add npm package build target

rzhw committed:
Make postcss@^7.0.7 a peerDependency

rzhw committed:
Use BEGIN-INTERNAL/END-INTERNAL to expose root BUILD in npm package

rzhw committed:
Document installation via npm or repository rules

nex3 commented on BUILD:
```diff
@@ -13,6 +13,9 @@
 # limitations under the License.
 
 load("@bazel_skylib//:bzl_library.bzl", "bzl_library")
+# BEGIN-INTERNAL
```
Should these be `BEGIN-`/`END-EXTERNAL`? Alternately, can we just remove them with Copybara without the guards?

rzhw replied:
In this case, these are comments that `pkg_npm` uses for removing npm-bound lines, rather than for Copybara.

nex3 reviewed (approved):
I'm not familiar with Bazel/npm integration but this looks reasonable to me. Is there someone else who's done something similar you can tag in for a secondary review?

alexeagle commented on package.json:
```diff
@@ -19,10 +19,13 @@
+    },
     "devDependencies": {
         "@bazel/typescript": "^1.0.1"
     },
```
bazelWorkspaces on the next line is also vestigial in rules_nodejs 2.0

rzhw replied:
Removed

alexeagle commented on README.md:
```diff
@@ -18,7 +18,65 @@ with Bazel.
+### Using npm and Node.js rules
+
+We recommend you use npm to install the PostCSS build rules if you use the
+[Node.js build rules][rules_nodejs], and **strongly recommend** using npm
```
does the 'Using repository rules' alternative below have any advantages? Why not give users a single install instruction?

rzhw replied:
I believe users may prefer the standard repository rules approach rather than being required to set up rules_nodejs, despite the advantages available for installing via npm.

alexeagle replied:
well I think rules_nodejs is pretty minimal, just an http_archive and a npm_install rule and no dependencies, but maybe there's a case for this, you know your own users.

I would point out this increases the surface area you need to test - ideally you'd want bazel-in-bazel integration tests to ensure the install and repository rules work following your instructions.

rzhw replied:
Good point - filed integration tests as #44.

alexeagle commented on README.md:
```diff
@@ -18,7 +18,65 @@ with Bazel.
+`package.json`.
+
+Your `WORKSPACE` should declare a `yarn_install` or `npm_install` rule named
+`npm`. It should then install the rules found in the npm packages using the
```
We have deprecated the install_bazel_dependencies function in rules_nodejs 2.0, due in a week or so.
The reason is that load()ing from `@npm` causes eager install from npm even for builds that don't request any frontend artifacts.

rzhw replied:
Removed

alexeagle commented on README.md:
```diff
@@ -18,7 +18,65 @@ with Bazel.
+
+# Setup repositories which are needed for PostCSS. This is an optional step
+# because you can always setup the required repositories yourself.
+load("@build_bazel_rules_postcss//:defs.bzl", "postcss_repositories")
```
why do you have transitive dependencies that aren't fetched from npm? can you avoid this? I don't see why postcss depends on sass or typescript, nor why those need bazel workspaces installed.
(the package.json here indicates that typescript should be a dev dependency and shouldn't leak into the distribution)
(also this //:defs.bzl is wrong, should be //:repositories.bzl)

rzhw replied:
This is known hence the README is a bit awkward, want to resolve this in #43.

rzhw replied:
On closer look this can just be removed now

alexeagle commented on README.md:
```diff
@@ -18,7 +18,65 @@ with Bazel.
+See https://github.com/bazelbuild/rules_nodejs/#quickstart
+
+This causes the `@bazel/postcss` package to be installed as a Bazel workspace
+named `build_bazel_rules_postcss`.
```
in 2.0 we now load just from `@npm//@bazel/postcss:index.bzl` there is no additional workspace

alexeagle replied:
2.0-rc.0 was shipped yesterday FYI

rzhw replied:
Thanks!

rzhw replied:
@alexeagle On upgrading this repo to 2.0.0 locally my test repo doesn't seem to work anymore:

    load("@npm//@bazel/postcss:defs.bzl", "postcss_binary", "postcss_plugin")

Gets me:

    ERROR: error loading package '': cannot load '@npm//@bazel/postcss:defs.bzl': no such file

See the [paste](https://gist.github.com/rzhw/d8217d54d74d23bc354789d72caf68a7). Any pointers for what could be going wrong?

I've run `yarn` successfully and can confirm my `node_modules` is populated as expected per
<img width="216" alt="Screen Shot 2020-07-03 at 3 50 57 pm" src="https://user-images.githubusercontent.com/79461/86436129-fdb2ca00-bd44-11ea-9d45-358e6365f5c7.png">

With the current state of this repo

rzhw replied:
Resolved.

alexeagle commented on README.md:
````diff
@@ -18,7 +18,65 @@ with Bazel.
+```python
+# Fetch transitive Bazel dependencies of PostCSS. This is an optional step
+# because you can always fetch the required transitive dependencies yourself.
+load("@build_bazel_rules_postcss//:package.bzl", "rules_postcss_dependencies")
````
same comment as below, can you drop this requirement? it causes eager install for builds that don't want to do css processing

rzhw replied:
This refers to https://github.com/bazelbuild/rules_postcss/pull/42#discussion_r441634677, right? If so let's resolve this issue in #43.

rzhw replied:
On closer look this can just be removed now

alexeagle commented on README.md:
```diff
@@ -18,7 +18,65 @@ with Bazel.
 ## Setup
 
-To be documented.
+### Using npm and Node.js rules
```
Feels like the README could address the tradeoffs between using these rules vs. just using postcss binary directly (I'll add a postcss example to https://github.com/bazelbuild/rules_nodejs/tree/1.x/examples/app/styles)

rzhw replied:
This makes me concerned whether taking `@bazel/postcss` might be breaking expectations of other users of packages generated from rules_nodejs - what do you think of e.g. forwarding the `npm_package_bin`-generated rule in addition to providing the `postcss_{binary,plugin}` rules

alexeagle replied:
I don't think there's a problem with one npm package exposing some custom rules around some tool, while the npm package for that tool can also be used standalone. We already have the same situation with `@bazel/typescript` (the g3 toolchain) vs just `typescript` which you can use straight from MS. For TS I documented tradeoffs https://github.com/bazelbuild/rules_nodejs/blob/master/docs/TypeScript.md#alternatives
so I'm just imagining a couple sentences similarly explaining what are the choices and why to pick one or the other.

rzhw replied:
Thanks! Added a TODO to the readme

alexeagle reviewed (commented):
Are you going to update release instructions? There are nuances around stamping which I'm happy to help with

rzhw committed:
rules_nodejs 2.0.0-rc.0

rzhw committed:
Code review

rzhw committed:
Move defs.bzl to index.bzl, add absolute paths that get substituted

rzhw committed:
Missed this s/defs.bzl/index.bzl change

rzhw committed:
Specify .bazelversion of 3.0.0

rzhw committed:
Switch ctx.actions.run to run_node as required for rules_nodejs 2.0

rzhw committed:
Update rules_sass

We need a version where install_source_map_support is removed i.e. at https://github.com/bazelbuild/rules_sass/commit/ff9c3677d6a330943812a8acd76b79aabc7c82cd

rzhw commented:
Status so far:

- [ ] Add release instructions
- [ ] Figure out why custom plugins are broken (e.g. `Error: Cannot find module 'build_bazel_rules_postcss/examples/custom_plugin/unquote.js'`)

rzhw committed:
rules_nodejs 2.0.0-rc.1

rzhw committed:
Create DEVELOPING.md

This is based off https://github.com/bazelbuild/rules_nodejs/blob/master/DEVELOPING.md, but reduced to only basic npm release instructions.

rzhw committed:
When using run_node, workspace paths must be resolved with the runfiles helper

rzhw committed:
rules_nodejs 2.0.0-rc.2

rzhw committed:
Use ctx.actions.run for wrapper

rzhw commented:
@alexeagle PTAL!

rzhw committed:
rules_nodejs 2.0.0-rc.3 now that bazelbuild/rules_nodejs#2023 is released

alexeagle commented:
LGTM, we should have rules_nodejs 2.0 final on Monday or Tuesday.

rzhw committed:
2.0.0 final

rzhw committed:
@bazel/typescript@^2.0.0

rzhw committed:
Mark as 0.5.0
